daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs

```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
                   then withNavigator
                   else (\_ _ _ f -> f sandboxPh)
 
-runDeploy :: String -> SandboxPort -> IO ()
-runDeploy host sandboxPort = do
+data HostAndPortFlags = HostAndPortFlags { hostM :: Maybe String, portM :: Maybe Int }
+
+withHostAndPortDefaults :: HostAndPortFlags -> IO HostAndPort
+withHostAndPortDefaults HostAndPortFlags{hostM,portM} = do
+    host <- fromMaybeM getProjectLedgerHost hostM
+    port <- fromMaybeM getProjectLedgerPort portM
+    return HostAndPort {host,port}
+
+runListParties :: HostAndPortFlags -> IO ()
+runListParties flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Listing parties at " <> show hp
+    xs <- Ledger.listParties hp
+    when (null xs) $ putStrLn "no parties are known"
+    mapM_ print xs
+    exitSuccess
+
+runAllocateParty :: HostAndPortFlags -> String -> IO ()
+runAllocateParty flags name = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Checking party allocation at " <> show hp
+    allocatePartyIfRequired hp name
+    exitSuccess
+
+runDeploy :: HostAndPortFlags -> IO ()
+runDeploy flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Deploying to " <> show hp
+    parties <- getProjectParties
+    mapM_ (allocatePartyIfRequired hp) parties
     darPath <- getDarPath
     doBuild
-    let SandboxPort port = sandboxPort
-    putStrLn $ "Deploying " <> darPath <> " to ledger on " <> host <> ":" <> show port
+    putStrLn $ "Uploading " <> darPath <> " to " <> show hp
     bytes <- BS.readFile darPath
-    let ls = L.uploadDarFile bytes
-    let timeout = 30 :: L.TimeoutSeconds
-    let ledgerClientConfig = L.configOfHostAndPort (L.Host $ fromString host) (L.Port port)
-    L.runLedgerService ls timeout ledgerClientConfig >>= \case
-        Right () -> do
-            putStrLn "Deploy succeeded."
-            exitSuccess
-        Left e -> do
-            hPutStrLn stderr $ "Deploy failed: " <> e
-            exitFailure
+    Ledger.uploadDarFile hp bytes
+    putStrLn "Deploy succeeded."
+    exitSuccess
+
+
+allocatePartyIfRequired :: HostAndPort -> String -> IO ()
+allocatePartyIfRequired hp name = do
+    party <-
+        Ledger.lookupParty hp name >>= \case
```

Comment:
ah, ocaml...

I think the more idiomatic phrasing would be:

```Haskell
partyM <- Ledger
party <- flip fromMaybeM partyM $ do
    putStrLn $ "Allocating party for '" <> name <> "' at " <> show hp
    Ledger.allocateParty hp name
```

Reply:
I wish we had `whenNothing :: Applicative f => Maybe a -> f a -> f a`. @ndmitchell would you accept that for the `extra` package?

Reply:
concern is that I'd expect whenNothing to be when (isNothing x) ... and this doesn't have that type. flip fromMaybeM seems quite nice here, without inventing any new names.

Reply:
Happy to change to Fran's suggestion

Reply:
depends what you idioms are :)
but I'll change it